dbms/src/Storages/DeltaMerge/LocalIndexerScheduler.h

```diff
@@ -0,0 +1,223 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Stopwatch.h>
+#include <Common/UniThreadPool.h>
+#include <Storages/DeltaMerge/LocalIndexerScheduler_fwd.h>
+#include <Storages/KVStore/Types.h>
+#include <Storages/Page/PageDefinesBase.h>
+
+#include <boost/functional/hash.hpp>
+#include <condition_variable>
+
+namespace DB::DM
+{
+
+// Note: this scheduler is global in the TiFlash instance.
+class LocalIndexerScheduler
+{
+public:
+    // The file id of the DMFile.
+    struct DMFileID
+    {
+        explicit DMFileID(PageIdU64 id_)
+            : id(id_)
+        {}
+        PageIdU64 id;
+    };
+    // The page id of the ColumnFileTiny.
+    struct ColumnFileTinyID
+    {
+        explicit ColumnFileTinyID(PageIdU64 id_)
+            : id(id_)
+        {}
+        PageIdU64 id;
+    };
+    using FileID = std::variant<DMFileID, ColumnFileTinyID>;
+
+    struct Task
+    {
+        // Note: The scheduler will try to schedule fairly according to keyspace_id and table_id.
+        const KeyspaceID keyspace_id;
+        const TableID table_id;
+
+        // The file id of the ColumnFileTiny or DMFile.
+        // Used for the scheduler to avoid concurrently adding index for the same file.
+        const std::vector<FileID> file_ids;
+
+        // Used for the scheduler to control the maximum requested memory usage.
+        const size_t request_memory;
+
+        // The actual index setup workload.
+        // The scheduler does not care about the workload.
+        ThreadPool::Job workload;
+    };
+
+    struct Options
+    {
+        size_t pool_size = 1;
+        size_t memory_limit = 0; // 0 = unlimited
+        bool auto_start = true;
+    };
+
+private:
+    struct InternalTask
+    {
+        const Task user_task;
+        Stopwatch created_at{};
+        Stopwatch scheduled_at{};
+    };
+
+    using InternalTaskPtr = std::shared_ptr<InternalTask>;
+
+public:
+    static LocalIndexerSchedulerPtr create(const Options & options)
+    {
+        return std::make_shared<LocalIndexerScheduler>(options);
+    }
+
+    explicit LocalIndexerScheduler(const Options & options);
+
+    ~LocalIndexerScheduler();
+
+    /**
+     * @brief Start the scheduler. In some tests we need to start scheduler
+     * after some tasks are pushed.
+     */
+    void start();
+
+    /**
+     * @brief Blocks until there is no tasks remaining in the queue and there is no running tasks.
+     * Should be only used in tests.
+     */
+    void waitForFinish();
+
+    /**
+     * @brief Push a task to the pool. The task may not be scheduled immediately.
```

Comment:
Add comments about that if the task request more memory than the pool can handle, this function will throw an exception